File: LeetCode-Java/Binary Tree Upside Down.java
/**
 * Definition for binary tree
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public TreeNode UpsideDownBinaryTree(TreeNode root) {
        if(root == null || root.left == null)
            return root;
        else{
            TreeNode originalLeft = root.left;
            TreeNode originalRight = root.right;
            root.left = null;
            root.right = null;
            return UpsideDownBinaryTreePart(originalLeft, root, originalRight);
        }
    }
    
    public TreeNode UpsideDownBinaryTreePart(TreeNode target, TreeNode root, TreeNode right){
        if(target.left == null){
            target.left = right;
            target.right = root;
            return target;
        }
        else{
            TreeNode newTarget = target.left;
            TreeNode originalRight = target.right;
            target.left = right;
            target.right = root;
            return UpsideDownBinaryTreePart(newTarget, target, originalRight);
        }
    }
}

File: Leetcode-cpp/binary_tree_upside_down.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    TreeNode* upsideDownBinaryTree(TreeNode* root) {
        stack<TreeNode*> record;
        while (root != nullptr) {
            record.push(root);
            root = root->left;
        }
        TreeNode* result = nullptr;
        TreeNode* preRoot = nullptr;
        while (!record.empty()) {
            TreeNode* current = record.top();
            record.pop();
            if (result == nullptr) {
                result = current;
            }
            if (preRoot != nullptr) {
                preRoot->left = current->right;
                preRoot->right = current;
            }
            preRoot = current;
            preRoot ->left = nullptr;
            preRoot ->right = nullptr;
        }
        
        return result;
        
    }
};